docs/architecture/domain.md

```diff
@@ -0,0 +1,116 @@
+
+ - `main` - Job that runs on each change to the source code.  _Note: this is also the same job that runs on proposed changes via GitHub Pull Requests_
+
+### Workflow
+
+Workflow is the defined order that [jobs] will execute in. Jobs can be executed in parallel, series, and a combination of the two to allow for all possibilities. All jobs executed in a given workflow run share:
+
+ - Source code checked out to the same git commit from the `main` [build] that triggered this
+ - Access to the [metadata] from the `main` [build] that triggered this
+
+There are two requirements for your pipeline's workflow:
+
+ - Must start with `main` (triggered when the git repository changes)
+ - Must contain every job (no non-triggered jobs)
+
+In the following example, this is the flow that will happen:
+```yaml
+workflow:
+    - main
+    - publish
+    - parallel:
+        - series:
+            - deploy-west
+            - validate-west
+        - series:
+            - deploy-east
+            - validate-east
+```
+
+ - `main` triggers `publish`
+ - `publish` triggers `deploy-west` and `deploy-east` in parallel
+ - `deploy-west` triggers `validate-west`
+ - `deploy-east` triggers `validate-east`
+
+### Job
+
+A job consists of executing multiple sequential [steps] inside a specified [container]. If any step in the series fails, then the entire job is considered failed and subsequent steps will be skipped (unless configured otherwise).
+
+Jobs share four pieces of context between steps:
+
+ - Source Code - Checked out to a given revision at the start of the job
+ - Filesystem - Any changes made by a step to the filesystem will be visible to subsequent steps
+ - Container - Any processes started or stopped by a step will be visible to subsequent steps
+ - Metadata - Structured key/value storage of relevant information
+
+Jobs are started by changes in source code, as a trigger from an upstream job, or by human decision.
+
+### Matrix
+
+A build matrix is a set of environment permutations for a given [job]. Permutations are usually used for testing against multiple [containers] or test types.
+
+In this example, the job will run 4 times for a given [build]:
+```yaml
+matrix:
+    NODE_VERSION:
+        - 4
+        - 6
+    EXPRESS_VERSION:
+        - 2
+        - 3
+```
+
+ - `NODE_VERSION=4` and `EXPRESS_VERSION=2`
+ - `NODE_VERSION=4` and `EXPRESS_VERSION=3`
+ - `NODE_VERSION=6` and `EXPRESS_VERSION=2`
+ - `NODE_VERSION=6` and `EXPRESS_VERSION=3`
+
+### Step
+
+A step is a named action that needs to be performed, usually a single shell command. If the command finishes with a non-zero exit code, the step is considered a failure.
+
+### Container
+
+A Container runs a [task] in an isolated environment. This is done in order to determine dependency version compatibility, without affecting other tasks that may be running at the some time. This is implemented using Docker containers.
```

Comment:
Regarding multiple versions of multiple dependencies being compatible with one another.  
This could be worded better. Thanks!